lib/zip_source_zip_new.c

```diff
@@ -80,10 +110,12 @@ zip_source_t *_zip_source_zip_new(zip_t *srcza, zip_uint64_t srcidx, zip_flags_t
     }
 
     partial_data = len < st.size;
-    needs_decrypt = ((flags & ZIP_FL_ENCRYPTED) == 0) && (st.encryption_method != ZIP_EM_NONE);
-    needs_decompress = ((flags & ZIP_FL_COMPRESSED) == 0) && (st.comp_method != ZIP_CM_STORE);
+    encrypted = (st.valid & ZIP_STAT_ENCRYPTION_METHOD) && (st.encryption_method != ZIP_EM_NONE);
+    needs_decrypt = ((flags & ZIP_FL_ENCRYPTED) == 0) && encrypted;
+    compressed = (st.valid & ZIP_STAT_COMP_METHOD) && (st.comp_method != ZIP_CM_STORE);
+    needs_decompress = ((flags & ZIP_FL_COMPRESSED) == 0) && compressed;
     /* when reading the whole file, check for CRC errors */
-    needs_crc = ((flags & ZIP_FL_COMPRESSED) == 0 || st.comp_method == ZIP_CM_STORE) && !partial_data;
+    needs_crc = ((flags & ZIP_FL_COMPRESSED) == 0 || !compressed) && !partial_data;
```

Comment:
When reopening, `needs_crc` should be false, since we don't have a CRC to compare it against. (`needs_crc` is used to compare the CRC from the actual data with the one recorded in the zip archive.)

Reply:
So the CRC source gets the checksum from underlying source by calling `zip_source_stat`, so I thought that if the custom source actually returns CRC in its stat object, why blocking the use of it?

Reply:
Yes, that makes sense. In that case add a `&& (st.valid & ZIP_STAT_CRC)` to `needs_crc` so it only gets added if we have a CRC to compare against.